### DISCUSSION TITLE
[OSPP 2022] project idea discussion

### DISCUSSION THREAD

**Discussion by Jave-Chen** [apache/dolphinscheduler#9723] [Ideas]
Hi, all, DS has participated in the OSPP 2022 event [1] and can publish 8 projects. 
If you have good ideas, you can post your ideas here for discussion. The idea includes 3 contents
- Project Description
- Project Output Standard
- Technical Requirements

Deadline May 10, 2022.
[1] https://mp.weixin.qq.com/s/3ROTqTSE9DlfClHkPQbwTw

**Comment by Jave-Chen** [comment]
Example for Basic Project

1. Add or adapt schedule management application for UKUI desktopof openEuler

- Project Description:
At present, a calendar is provided in the lower right corner of the UKUI desktoptaskbar, but it is unable to manage the schedule and does not improvetheoverall ecology. We hope to develop or adapt a schedule management relatedsoftware for the UKUI desktop.

- Project Output Standard
1. Be able to add, delete and edit the schedule. 
2. It can display the schedule in a variety of views, such as month, weekandday. 
3. Support taskbar tray. 
4. It's better to synchronize the schedule established by outlook or other
accounts. 

- Technical Requirements:
Familiar with Linux operating system, with research and development abilityunder Linux system. 

2. Provide scalable application visualization configuration softwareinopenEuler system.

- Project Description:
For a long time, Linux needs professional technicians to operate and manage,
the technical threshold is high, and it is not friendly for beginners. Linuxenvironment has excellent software package management mechanism, but inthe use of software, File-based configuration also brings the tediousnessof
manual editing. Compared with windows application ecology, all softwarehasgood configuration visualization function. Our goal is to reduce the threshold of using Linux environment and enablebeginners to quickly start using Linux environment. 

- Project Output Standard
1. Complete the design, development and testing of the project, and providethe documents of architecture, use and plug-in secondary development. 
2. The front end provides as detailed a configuration description as possible. 
3. Give priority to designing core functions as independent libraries, suchasplug-in libraries, etc. 
4. Support the configuration of at least five common application software(suchas mysql, redis, httpd, etc.), and consider the configuration differencesbetween different versions, the built-in application configuration of the system(such as network, firewall, crontab, etc.)can also be extended to providetheeffective function of restart service. 
5. Open source in openEuler community.

- Technical Requirements:
1. Use go language for development. 
2. Front-end visualization and configuration based on Web. 
3. Requires a plug-in architecture to support the configuration of any extendedsoftware

---
基础类项目示例
1. 为 openEuler 的 UKUI 桌面增加或适配日程管理应用
- 描述：
目前在 UKUI 桌面任务栏的右下角，提供了日历，但是却无法管理日程，未完善整体生态，希望能够针对 UKUI 桌面研发或者适配一款日程管理的相关软件。
- 产出标准：
1. 能够增加、删除、编辑日程。
2. 能够多种视图来显示日程，如月、周、日的模式显示日程。
3. 支持任务栏托盘。
4. 最好能够同步 outlook 或是其它账户建立的日程。
- 技术要求：
熟悉 linux 操作系统，具备在 Linux 系统下的研发能力

2. 在 openEuler 系统中提供可扩展的应用可视化配置软件
- 项目描述
linux 长期以来需要专业的技术人员进行操作及管理，技术门槛较高，对于初学者使用不够友好。linux 环境已经拥有优秀的软件包管理机制，但是在软件使用上，基于文件的配置也带来手动编辑的繁琐。对比 windows 应用生态，所有的软件都具有良好的配置可视化功能。我们的目标是：降低 linux 环境的使用门槛，使初学者能快速上手使用linux 环境。
- 产出标准：
1. 完成项目的设计、开发、测试并提供架构和使用以及插件二次开发的文档2. 前端提供尽可能详细的配置描述
3. 优先考虑将核心功能设计成独立的库，如插件库等
4. 支持至少 5 个常用应用软件的配置（如 mysql、redis、httpd 等），并考虑不同版本之间的5. 配置差异，也可考虑扩展系统内置的应用配置（如网络、防火墙、crontab 等），提供重启服务生效功能
5. 在 openEuler 社区中开源
- 技术要求：
1. 使用 go 语言进行开发
2. 基于 web 的前端可视化及配置
3. 插件架构，支持扩展任意软件的配置

**Comment by Jave-Chen** [comment]
Example for Advanced Project

1. Provide OpenStack containerized deployment capabilitiesforopenEuler

- Project description:
    OpenStack is an open source cloud computing management platformproject, which is a combination of a series of open source software projects. It isalsoan open source code project authorized by the Apache License (afreesoftware license issued by the Apache Software Foundation), which was jointlydeveloped and initiated by NASA (National Aeronautics and SpaceAdministration) and Rackspace. OpenStack provides scalable, flexiblecloudcomputing services for both private and public clouds. The goal of the project
    is to provide a cloud computing management platform that is simpletoimplement, scalable, rich and standardized. The entire OpenStack frameworkconsists of a number of components that provide different services for cloudcomputing platforms. OpenStack contains many components, anditsinstallation deployment is a very headache problem. Using containerizationtechnology can significantly reduce the difficulty of deploying OpenStack. This topic uses the "OpenStack+Kubernetes+Isula" technology stack to realizethe rapid containerized deployment capability of OpenStack on openEuler.  

- Project Output Standard:
1. Complete the kubernetes deployment file writing correspondingtoOpenStack;
2. Deploy, test and verify the "OpenStack+kubernetes+isula" cluster
    environment, and output relevant test reports.

- Technical Requirements:
1. Familiar with Linux operating system. 
2. Understand basic cloud computing concepts. 
3. Understand and use container services such as docker or isula. 
4. Understand and use the kubernetes container orchestration service. 
5. Have some knowledge of virtualized or containerized networking.   

---

2. Optimization of Startup Times of StratoVirt Virtualization Platform

- Project description:
    Lightweight Virtual Machine (Mirto VM) is a new virtualization technologyinrecent years. StratoVirt is open source lightweight virtualization softwarefor
    Huawei openEuler. It is written in Rust, a memory security language. It isbased on Rust, and has the characteristics of complete functions, fast startup, and low memory overhead. StratoVirt has a wide application spaceincontainers and serverless scenarios. The aim of this experiment is to further
    improve StratoVirt's lightweight and low noise capabilities based on its existingcapabilities.
- Project Output Standard

1. Startup time to reach or close to Kunpeng 920 platforms on the X86_64platform. 
2. Summarize the entire development process and problems encountered, andform a document. 

- Technical Requirements:

1. Familiar with the Linux kernel boot process. 
2. Have a certain understanding of virtualization technology. 
3. Familiar with Rust. This topic requires optimization of the startup time of lightweight virtual
    machines on the x86_64 platform. You can try and are not limited tothefollowing solutions:
     - Optimize the bootloader module for StratoVirt. 
     - Clip guest kernel adapted to MicroVm. 
     - Startup by way of PVH.

---

进阶类项目示例

1. 为 openEuler 提供 OpenStack 容器化部署能力

- 描述：
    OpenStack 是一个开源的云计算管理平台项目，是一系列软件开源项目的组合。由NASA(美国国家航空航天局)和 Rackspace 合作研发并发起，以 Apache 许可证（Apache 软件基金会发布的一个自由软件许可证）授权的开源代码项目。OpenStack 为私有云和公有云提供可扩展的弹性的云计算服务。项目目标是提供实施简单、可大规模扩展、丰富、标准统一的云计算管理平台。整个 OpenStack 框架由众多组件组成，这些组件为云计算平台提供不同的服务。OpenStack 包含组件繁多，其安装部署是非常令人头疼的问题。使用容器化技术，可以显著降低 OpenStack 的部署难度。本题目使用 OpenStack+Kubernetes+Isula 技术栈，实现OpenStack在 openEuler 上的容器化快速部署能力。
- 产出标准：
    完成 OpenStack 对应的 kubernetes 部署文件编写；
    部署、测试并验证 OpenStack+kubernetes+isula 集群环境，输出相关测试报告。
- 技术要求：
    1、熟练使用 Linux 操作系统。
    2、了解基本的云计算概念。
    3、了解并会使用 docker 或 isula 等容器服务。
    4、了解并会使用 kubernetes 容器编排服务。
    5、对虚拟化或容器化网络有一定了解。

--- 

2. StratoVirt 虚拟化平台启动时间的优化

- 描述
    轻量虚拟机（Mirto VM）是近几年新兴的一项虚拟化技术，StratoVirt 是华为openEuler 开源的轻量虚拟化软件，使用内存安全语言 Rust 编写，基于 Rust 语言编写，具备功能完善、启动快速、内存开销小等特点，在容器和 serverless 场景有广阔的应用空间。本实验的目标是基于 StratoVirt 现有的能力，进一步提升其的轻量低噪的能力。
- 产出标准：

1. 在 x86_64 平台上达到或接近 kunpeng920 平台的启动时间
2. 总结整个开发流程和遇到的问题，并形成文档

- 技术要求：

1. 熟悉 Linux 内核启动流程
2. 对虚拟化技术有一定了解
3. 熟悉 Rust 语言
    本题目要求对轻量虚拟机在 x86_64 平台上的启动时长进行优化，可以尝试并不限于以下方案：
     - 优化 StratoVirt 的 bootloader 模块
     - 裁剪与 MicroVm 适配的 guest kernel 内核
     - 实现以 PVH 的方式启动

**Comment by SbloodyS** [comment]
## Resource center optimization

- Difficulty: Advanced

1. Refactoring the reading and writing logic of dolphinscheduler resource center
- Project Description
At present, the resource center reads and writes HDFS/S3/local directories and files through ```t_ds_resources``` record file/folder to realize file recording. However, when the user initializes dolphinscheduler or the user modifies the content stored by the third party outside the ds, the ds cannot synchronize in real time. Hope to ensure real-time synchronization with the content stored by the third party (call the API of the third party to add, delete, modify and check the file)

- Project Output Standard
1. DS does not store the path of the file in the resource center. The reading and writing of the resource center are all realized by calling the third-party API. Only the content path that the task instance depends on is saved to database for file deletion verification
2. Realize all the functions of the existing resource center. Including file/folder addition, deletion, modification and query, and task instance calling the resource center.
3. Backend API documentation

- Technical Requirements
1. Familiar with Linux file management
2. Understand HDFS/S3 components and be able to quickly learn the use of API according to official documents
3. Familiar with Java language

---

## 资源中心优化

- 难度：进阶

1. 重构 Dolphinscheduler 资源中心的读写逻辑
- 描述
目前资源中心读写 HDFS/S3/本地目录 通过在数据库表中 ```t_ds_resources``` 记录文件/文件夹目录来实现文件记录，但是在用户初始化DS或者用户在DS的外部修改了第三方存储的内容的时候DS无法进行实时同步，希望能够与第三方存储的内容保证实时同步（调用第三方的API对文件增删改查）

- 产出标准
1. DS 不存储资源中心中文件的路径，对于资源中心的读写全都通过调用第三方API实现，只保存任务实例依赖到的内容路径于数据库用于文件删除校验
2. 实现现有的资源中心的所有功能，其中包括 文件/文件夹 增删改查，任务实例调用资源中心
3. 后端API文档

- 技术要求
1. 了解 Linux 系统，对 Linux 文件管理有一定的熟悉
2. 了解 HDFS/S3 组件，能够较快速的根据官方文档学习API的使用
3. 熟悉 Java 语言

- 从中学到什么
我们希望每个加入 SoC 的人都能从项目中学到一些东西。当你完成这个项目时，你可以学到：

1. 更加熟悉开源社区、它是什么、它是如何工作的以及如何做出贡献
2. 如何10分钟内快速搭建简易的HDFS/S3环境
3. 了解HDFS/S3在实际生产环境中的作用
4. 了解关于大数据任务调度系统的知识，它是什么，它的重点是什么，它是如何运行的

**Comment by brave-lee** [comment]
Master/Worker/ API log optimization

- Description:
 because the DS log is inconvenient for real location, the printed log needs to be standardized and optimized
- Output criteria:
1. Sort out the DS printing log specification guidance document and submit it to the document library
2. Implement the master / worker / API module according to the specification document and submit it to the development branch
- Technical Requirements
Familiar with Java language
---------------------------------------------------

master/worker/api模块日志优化

描述：由于DS日志不方便真实定位问题，需要将打印日志进行规范及优化
产出标准：
1、整理出DS打印日志规范指导文档，并提交到文档库
2、将master/worker/api模块按照规范文档进行落地并提交到开发分支
 技术要求
熟悉 Java 语言

**Comment by zhongjiajie** [comment]
## Topic

Add file reader plugin for DolphinScheduler Python API

### label

* Support Language: English only
* Difficulty: Advanced
* Technical Field: IO, network, workflow
* Language: Python

## Description

Currently, DolphinScheduler Python API only could pass string format arg as tasks' parameter, such as our tutorial 
https://github.com/apache/dolphinscheduler/blob/49a1270ef5efd9e47d476752ed2e2e92a6ef99f1/dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/examples/tutorial.py#L52-L55
we only pass a string to `shell` task type. But generally, users hope the command is stored in other resources(for easier maintenance and integration), such as local file systems, GitHub, GitLab, Amazon S3, Aliyun OSS and etc. Maybe we could add trick like 

```py
@Resource(
    type="github",
    id="https://github.com/apache/dolphinscheduler/tree/dev/script"
)
with ProcessDefinition(
    name="tutorial",
    schedule="0 0 0 * * ? *",
    start_time="2021-01-01",
    tenant="tenant_exists",
) as pd:
    task_parent = Shell(name="task_parent", command="install.sh")
    task_child_one = Shell(name="task_child_one", command="start-all.sh'")
    task_child_two = Shell(name="task_child_two", command="stop-all.sh")
    task_union = Shell(name="task_union", command="status-all.sh")
```

And task task_parent will load file content from URL https://github.com/apache/dolphinscheduler/blob/dev/script/install.sh and pass it to parameter `command`, it makes our DAG file easier and maintainable

## Output Standard

* Add a plugin for `Resource`
* Support `Resource` of local file systems, GitHub, GitLab, Amazon S3, Aliyun OSS
* Pass tests coverage for Python API, the current threshold is 90% 
* Documentation, including development(how to implement plugin) and usage(each `Resource` plugin usage)

## Requirements

* Familiar with python, including with both decorator, io, network, unit tests, document

## What Can You Learn

We wish everyone joining SoC could learn some things from the project. When you finish this project, you could learn:

* Become more familiar with the open source community, what is it, how it work, and how to contributing
* How to write production-level Python codes and docs, you could improve your Python skill, how to write tests with `pytest` and `tox`, how to write a document with `sphnix` and it related plugin, how to format your Python code and the linter inside
* Adding knowledge about task scheduling system, what is it and what it focuses, how it could be run

---

## 课题

为 DolphinScheduler Python API 添加资源文件加载插件

### 标签

* 支持语言: 仅英文
* 难度: 进阶
* 技术范畴: IO, network, workflow
* 语言: Python

## 描述

目前，DolphinScheduler Python API 只能传递字符串格式传递任务的参数，例如我们的 tutorial
https://github.com/apache/dolphinscheduler/blob/49a1270ef5efd9e47d476752ed2e2e92a6ef99f1/dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/examples/tutorial.py#L52-L55
我们只将一个字符串传递给 `shell` 任务类型。但通常来说，用户希望具体的执行代码储存在其他系统中（希望更好维护和集成），例如将具体文件存储在本地文件系统、GitHub、GitLab、Amazon S3、阿里云 OSS 等。也许我们可以添加语法糖让用户更加简单的操作外部文件

```py
@Resource(
    type="github",
    id="https://github.com/apache/dolphinscheduler/tree/dev/script"
)
with ProcessDefinition(
    name="tutorial",
    schedule="0 0 0 * * ? *",
    start_time="2021-01-01",
    tenant="tenant_exists",
) as pd:
    task_parent = Shell(name="task_parent", command="install.sh")
    task_child_one = Shell(name="task_child_one", command="start-all.sh'")
    task_child_two = Shell(name="task_child_two", command="stop-all.sh")
    task_union = Shell(name="task_union", command="status-all.sh")
```

任务 task_parent 将从 URL https://github.com/apache/dolphinscheduler/blob/dev/script/install.sh 加载文件内容并将其传递给参数 `command`，它使我们的 DAG 文件更容易和可维护

## 产出

* 实现 `Resource` 的插件化
* 实现具体的插件 `Resource` 包括但不限于本地文件系统，GitHub, GitLab, Amazon S3, 阿里云 OSS
* Python API 的通过测试覆盖率，当前阈值为 90%
* 文档，包括开发（如何二开实现插件）和使用（每个 `Resource` 插件使用）

## 对报名者要求

* 熟悉python，包括decorator、io、network、unittest、document

## 从中学到什么

我们希望每个加入 SoC 的人都能从项目中学到一些东西。当你完成这个项目时，你可以学到：

* 更加熟悉开源社区、它是什么、它是如何工作的以及如何做出贡献
* 如何编写生产级 Python 代码和文档，你可以提高你的 Python 技能，如何使用 `pytest` 和 `tox` 编写测试，如何使用 `sphnix` 及其相关插件编写文档，如何格式化你的 Python 代码和里面的 linter
* 添加关于任务调度系统的知识，它是什么，它的重点是什么，它是如何运行的

**Comment by gabry-lab** [comment]
## Topic

detect and store all of application ids of yarn jobs automatically, instead of extracting them in task logs with a regexp

### Label

* Support Language: English only
* Difficulty: Basic
* Technical Field: Yarn, Hive, Spark SQL, Aspect-Oriented Programming
* Language: Java

## Description

DolphinScheduler support a wide spectrum of tasks, such as python, shell, etc. A basic function is to collect and kill all of the yarn jobs submitted by these tasks. 
For now, DolphinScheduler uses a regular expression to extract these yarn application id from task logs. I believe it is not a ideal way, and you are so clever to detect and store all of application ids of yarn jobs automatically.
Tips: on option is using AOP to achieve that.


## Output Standard

* detect and store all of application ids of yarn jobs automatically.
* the mechanism should be transparent to end users and developers
* 

## Requirements

* Familiar with Java, Yarn, Hive, Spark SQL, Aspect-Oriented Programming

## What Can You Learn

We wish everyone joining SoC could learn some things from the project. When you finish this project, you could learn Yarn, Hive, Spark SQL, Aspect-Oriented Programming


---

## 课题

自动收集、保存作业提交的Yarn作业的application id

### 标签

* 支持语言: 仅英文
* 难度: 进阶
* 技术范畴: Yarn, Hive, Spark SQL, Aspect-Oriented Programming
* 语言: Java

## 描述

DolphinScheduler支持各种各样的作业，比如Python、Shell等，其中一个很基础的功能就是收集这些作业创建的Yarn Job的application id。现在的实现方式是用一个正则表达式从作业的日志中解析，但这并不是一个完美的实现方式，本课题的任务就是用一种新的方式去自动收集、报错Yarn Job的application id信息，当然了，必须对用户透明，即对用户创建的作业无侵入。

## 产出

* 自动收集、报错Yarn Job的application id信息
* 必须对用户透明，即对用户创建的作业无侵入

## 对报名者要求

* 熟悉 Java, Yarn, Hive, Spark SQL, AOP

## 从中学到什么

我们希望每个加入 SoC 的人都能从项目中学到一些东西。当你完成这个项目时，你可以学到：

* Java, Yarn, Hive, Spark SQL, AOP的高级知识点

**Reply by liunaijie** [reply]
@gabrywu Hi, May I ask is this feature implemented now?
This is a good feature, we want use this feature to build relactionship between DS workflow with Yarn Job.

**Reply by gabry-lab** [reply]
Yes

**Comment by caishunfeng** [comment]
## Topic

Add workflow demos for DolphinScheduler server initialization.

### Label

- Supported language: English
- Difficulty: Basic
- Technical Field: Spring, workflow, Shell
- Language: Java, Shell

## Description

After users start the DolphinScheduler service, they can use the demo-tool program to preset workflow demos,, including but not limited to simple shell tasks, logical tasks (switch, dependent, subprocess, condition), parameter passing and other functions, which can guide users more easily use DolphinScheduler.

## Output Standard

- Tool program for preset workflow demo
- Add init mode to service startup script;
- Workflow demos include but are not limited to shell tasks, logical component tasks (switch, dependent, subprocess, condition), parameter passing, etc.;

## Requirements

- Familiar with Java、Shell

## What can we learn

We hope everyone who joins SoC will learn something from the project. When you complete this project, you can learn:

- Become more familiar with the open source community, what it is, how it works and how to contribute;
- Familiar with the `apache/dolphinscheduler` project, knowledge about the task scheduling system, what is its focus, and how it works;



---



## 课题

为 DolphinScheduler 添加服务初始化的工作流demo

### 标签

- 支持语言: 仅英文
- 难度: 中等
- 技术范畴: Spring、workflow、Shell
- 语言: Java、Shell

## 描述

用户在启动DolphinScheduler服务后，可以使用demo-tool程序预置工作流demo，包括但不限于简单的shell任务、逻辑组件任务（switch、dependent、subprocess、condition）、参数传递等功能， 引导用户更方便地使用DolphinScheduler。

## 产出

- 预置工作流demo的tool程序；
- 服务启动脚本添加 init模式；
- 工作流demo包括但不限于shell任务、逻辑组件任务（switch、dependent、subprocess、condition）、参数传递等；

## 对报名者要求

- 熟悉Java、Shell

## 从中学到什么

我们希望每个加入 SoC 的人都能从项目中学到一些东西。当你完成这个项目时，你可以学到：

- 更加熟悉开源社区、它是什么、它是如何工作的以及如何做出贡献；
- 熟悉 apache/dolphinscheduler 项目，关于任务调度系统的知识，它的重点是什么，它是如何运行的；

**Comment by zhuangchong** [comment]
## Topic

Add Java task type for DolphinScheduler

### Label

- Supported language: English
- Difficulty: Advanced
- Technical Field: Spring, workflow, Shell
- Language: Java、Shell、TypeScript

## Description

Provides Java type scheduling tasks, including Jar and custom Java code, and defines the Resources, Libs resource files, and JVM parameters required by Java runtime through the WEBUI, cover timed task calls Java program scenarios and enhance user-friendly configuration.

## Output Standard

- Executable Java type task plugin;
- The Java type plug-in supports the Jar method and provides dynamic parameter input of Resources, Libs resource files and JVM parameters;
- Java type plug-in supports WebUI custom Java code method, provides Java class instant compilation and execution;

## Requirements

- Familiar with Shell、Java、SPI、Java compiler、VUE3、TypeScript

## What can we learn

We hope everyone who joins SoC will learn something from the project. When you complete this project, you can learn:

- Become more familiar with the open source community, what it is, how it works and how to contribute;
- Familiar with the `apache/dolphinscheduler` project, knowledge about the task scheduling system, what is its focus, and how it works;
- Understand the discovery mechanism provided by the SPI service to improve program scalability;


---



## 课题

为 DolphinScheduler 添加 Java 任务类型

### 标签

- 支持语言: 仅英文
- 难度: 进阶
- 技术范畴: Spring、workflow、Shell
- 语言: Java、Shell、TypeScript

## 描述

提供Java类型的调度任务，包括Jar和自定义Java代码两种方式，通过WEBUI定义Java运行时所需要的Resources、Libs资源文件和JVM参数，覆盖定时任务调用Java程序场景并提升用户方便配置。

## 产出

- 可执行的Java类型任务插件;
- Java类型插件支持Jar方式，提供Resources、Libs资源文件和JVM参数的动态参数输入；
- Java类型插件支持WebUI自定义Java代码方式，提供Java类即时编译，执行；

## 对报名者要求

- 熟悉Shell、Java、SPI、Java compiler、VUE3、TypeScript

## 从中学到什么

我们希望每个加入 SoC 的人都能从项目中学到一些东西。当你完成这个项目时，你可以学到：

- 更加熟悉开源社区、它是什么、它是如何工作的以及如何做出贡献；
- 熟悉 apache/dolphinscheduler 项目，关于任务调度系统的知识，它的重点是什么，它是如何运行的；
- 了解SPI服务提供发现机制，提升程序可扩展性；

**Comment by wen-hemin** [comment]
## Topic

Support test tasks

### Label

- Supported language: English
- Difficulty: Advanced
- Technical Field: VueJs、Spring
- Language: Java、SQL

## Description

At present, it is difficult to verify whether the workflow configuration is correct before the new/modified workflow goes live, and can only be verified by executing it in the online environment, so it is necessary to support the test environment to execute tasks.
The data source center supports the configuration of test environment sources, and online environment sources can be bound to a test environment source. Task execution and workflow execution support the option "Whether to test". When “test” is selected, the Worker node automatically replaces the online environment source with the bound test environment source before the task is executed, and then cooperates with the Worker group to realize the isolation of online and test environments. Finally achieve the goal of executing the test.

## Output Standard

- Added online and test options to the data source configuration page, and supports binding test sources
- Added online and test options on workflow execution page and task execution page
- Master node and Worker node parse the execution environment identifier and replace the data source with the test data source
- Added execution environment identifiers to workflow instance and task instance pages

## Requirements

- Familiar with Java、Spring、VUE3

## What can we learn

We hope everyone who joins SoC will learn something from the project. When you complete this project, you can learn:

- Become more familiar with the open source community, what it is, how it works and how to contribute;
- Familiar with the `apache/dolphinscheduler` project, knowledge about the task scheduling system, what is its focus, and how it works;


---



## 课题

支持测试任务

### 标签

- 支持语言: 仅英文
- 难度: 进阶
- 技术范畴: VueJs、Spring
- 语言: Java、SQL

## 描述

目前新增/修改的工作流在上线前很难验证工作流配置是否正确，只能通过在线上环境执行来验证，所以有必要支持测试环境执行任务。
数据源中心支持配置测试环境源，线上环境源可绑定一个测试环境源。任务执行及工作流执行支持“是否测试”选项，选择测试时，Worker节点在任务执行前，自动替换线上环境源为绑定的测试环境源，再配合Worker分组实现线上、测试环境的隔离，最终达到支持测试任务的目标。

## 产出

- 数据源配置页面增加线上、测试选项，支持绑定测试源
- 工作流执行页面、任务执行页面新增线上、测试选项
- Master节点、Worker节点解析执行环境标识，替换数据源为测试数据源
- 工作流实例、任务实例页面增加执行环境标识

## 对报名者要求

- 熟悉Java、Spring、VUE3

## 从中学到什么

我们希望每个加入 SoC 的人都能从项目中学到一些东西。当你完成这个项目时，你可以学到：

- 更加熟悉开源社区、它是什么、它是如何工作的以及如何做出贡献；
- 熟悉 apache/dolphinscheduler 项目，关于任务调度系统的知识，它的重点是什么，它是如何运行的；

**Comment by zixi0825** [comment]
## Topic
Add authorization management of read and write permissions for DolphinScheduler's project and resource center

## Label
Supported language: English
Difficulty: Simple
Technical Field: VueJs、Spring、TypeScript
Language: Java、SQL

## Description
At present, DolphinScheduler's control of permissions is relatively simple, and it is hoped to increase the control of operation permissions to separate read and write permissions when authorizing projects and resource centers.

## Output Standard
- Distinguish read and write permissions for the entire content of a single project, and grant read-only or read-write permissions to a single user for a single project
- Distinguish read and write permissions for all content in the resource center, and distinguish read-only or read and write permissions based on the authorized content of the current resource center
- Add read-only and read-write authorization operations to the front-end page when authorization is performed

## Requirements
Familiar with Java、Spring、Vue3、TypeScript

## What can we learn
We hope everyone who joins SoC will learn something from the project. When you complete this project, you can learn:

- Become more familiar with the open source community, what it is, how it works and how to contribute;
- Familiar with the apache/dolphinscheduler project, knowledge about the task scheduling system, what is its focus, and how it works;
- Learn some knowledge about permission management

## 课题
为 DolphinScheduler 的项目和资源中心增加读写权限的授权管理 

## 标签
- 支持语言: 仅英文
- 难度: 简单
- 技术范畴: Vue3、Spring、TypeScript
- 语言: Java

## 描述
当前DolphinScheduler对于权限的管控相对比较简单，希望在进行项目和资源中心授权时增加操作权限管控来进行读写权限的分离。

## 产出
- 针对单个项目的全部内容进行读写权限区分，对单个用户针对单个项目授权只读或者读写权限
- 针对资源中心的全部内容进行读写权限区分，基于当前资源中心的授权内容区分出只读或者读写权限
- 进行授权时前端页面增加只读和读写的授权操作

## 对报名者要求
熟悉Java、Spring、Vue3、TypeScript

## 从中学到什么
我们希望每个加入 SoC 的人都能从项目中学到一些东西。当你完成这个项目时，你可以学到：

- 更加熟悉开源社区、它是什么、它是如何工作的以及如何做出贡献；
- 熟悉 apache/dolphinscheduler 项目，关于任务调度系统的知识，它的重点是什么，它是如何运行的；
- 学习权限管理相关知识
